src/protocol/dot_product.rs

```diff
@@ -0,0 +1,173 @@
+/// an instantiation for Replciated Secret Sharing is provided in section 6.1 on page 27
+///
+/// To summarize the CHIKP protocol:
+/// each helper computes `u_i+1 = (s_i + s_i+1) * (t_i + t_i+1) - s_i+1 * t_i+1`
+/// It then randomizes this share with randomness that it shares with the
+/// helper to its left, and sends it to the helper to its right,
+/// and receives `u_i` from the helper to its left.
+/// Each helper then defines the pair `(u_i, u_i+1)` as its shares of the output.
+///
+/// (Note it is equivalent to compute `u_i+1 = s_i * t_i + s_i+1 * t_i + s_i * t_i+1`,
+/// but this requires three multiplications instead of two. Assuming multiplying field elements is
+/// significantly more costly (in CPU terms) than addition, this is less efficient to compute)
+///
+/// The key observation that makes this protocol work, is that each helper can compute many `u_i+1` values
+/// from many multiplications, sum them all together, randomize them with correlated randomness, and can
+/// send this single field value to the next helper. The result will be a replicated secret sharing of the
+/// dot-product of the two vectors.
+///
+pub mod stream {
+    use crate::{
+        error::BoxError,
+        field::Field,
+        helpers::{fabric::Network, Direction},
+        protocol::{context::ProtocolContext, RecordId},
+        secret_sharing::Replicated,
+    };
+    use futures::stream::StreamExt; // for `next`
+    use futures::Stream;
+    use std::pin::Pin;
+
+    use serde::{Deserialize, Serialize};
+
+    /// A message sent by each helper when they've computed one share of the result
+    #[derive(Debug, Serialize, Deserialize, Default)]
+    pub struct UValue<F> {
+        u: F,
+    }
+
+    type PairOfShares<F> = (Replicated<F>, Replicated<F>);
+
+    ///
+    /// Consumes the input stream of pairs of replicated secret shares and
+    /// for each pair, multiplies them, and accumulates the result in a running sum.
+    /// once all the values in the stream are consumed, it randomizes the sum with
+    /// randomness it shares with the helper to the left, and sends to the right.
+    /// It receives a randomized sum from the helper to its left, and is now in
+    /// possession of a replicated secret sharing of the dot-product of the two
+    /// streaming vectors of replicated secret shares.
+    ///
+    /// ## Errors
+    /// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+    /// back via the error response
+    #[allow(dead_code)]
+    pub async fn accumulate_dot_product<F: Field, N: Network>(
+        ctx: ProtocolContext<'_, N>,
+        record_id: RecordId,
+        mut input_stream: Pin<Box<dyn Stream<Item = PairOfShares<F>>>>,
+    ) -> Result<Replicated<F>, BoxError> {
+        let prss = &ctx.prss();
+        let (s_left, s_right): (F, F) = prss.generate_fields(record_id);
+
+        let mut u_right = s_right - s_left;
+
+        while let Some(item) = input_stream.next().await {
```

Comment:
how do you plan to implement check-pointing? I would imagine with this model you need to open a new stream inside every maliciously secure protocol and send everything to that stream. There is only one place where you can do the verification which is after this function returns. I wonder if you would want intermediate checkpoints as well - that may require incorporating verification and dot product together

Reply:
I imagine that you'll need to say:

let v = MaliciousSecurityValidator::new();

And then you'll need to tell it about all of the inputs that go into the circuit, and every time you encounter a multiplication, or a reshare, you'll need to tell it about that. Then, at the end of some period of time (e.g. before we reveal a permutation within our sort protocol) you'll say:

v.validate();

It'll close out the two streams, which will finalise the computation of these two dot products ( $[w]$ and $[u]$), reveal $r$, and finally call `CheckZero` on $r[w] - [u]$. 

If it's zero, all good. If it's NOT zero, abort the entire protocol and file an incident report.